oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedSortBatchTask.java

```diff
@@ -116,48 +136,90 @@ public Result call() throws Exception {
         }
     }
 
+    private void buildSortArray(NodeStateEntryBatch nseb) {
+        Stopwatch startTime = Stopwatch.createStarted();
+        ByteBuffer buffer = nseb.getBuffer();
+        int totalPathSize = 0;
+        while (buffer.hasRemaining()) {
+            int positionInBuffer = buffer.position();
+            // Read the next key from the buffer
+            int pathLength = buffer.getInt();
+            totalPathSize += pathLength;
+            if (pathLength > copyBuffer.length) {
+                LOG.debug("Resizing copy buffer from {} to {}", copyBuffer.length, pathLength);
+                copyBuffer = new byte[pathLength];
+            }
+            buffer.get(copyBuffer, 0, pathLength);
+            // Skip the json
+            int entryLength = buffer.getInt();
+            buffer.position(buffer.position() + entryLength);
+            // Create the sort key
+            String path = new String(copyBuffer, 0, pathLength, StandardCharsets.UTF_8);
+            String[] pathSegments = SortKey.genSortKeyPathElements(path);
```

Comment:
Just an idea: if we don't need "path" and only the array, then maybe we could store the number of entries and then the entries separately?

Reply:
Something like that:

    int pathSegmentCount = buffer.getInt();
    String[] pathSegments = new String[pathSegments];
    for (int i = 0; i < pathSegmentCount; i++) {
        pathSegments[i] = readString(buffer);
    }

this is with a readString utility method, which maybe would anyway make sense.

Reply:
I don't think this would be faster. I tried a similar idea when optimizing the merge phase, when reading lines from the intermediate sorted files. It was about the same speed to read the segments of the path directly one by one than to read the full path as a string and then break it into segments, even though in the second option we are allocating an extra String. I was very surprised. I think the reason is that with the path as a String, we can use the `String.indexOf` method, which is an intrinsic and likely heavily optimized. If we read the path segments from the byte buffer, we have to iterate byte by byte in the buffer until finding the `/` characters, then build a string. This will very likely be slower than just reading the bytes with the path in a single operation from the byte buffer and then breaking the path string into segments. Furthermore, the String with the path is very short lived, so the JIT is likely applying escape analysis to allocate it on the stack.

Reply:
> if we read the path segments from the byte buffer, we have to iterate byte by byte in the buffer until finding the / characters

No. The above code first writes the number of path segments, and then each path segment. There is no search for '/'.

Reply:
Maybe I should also add readString:

    private static String readString(ByteBuffer buff) {
        int len = buff.getShort();
        if (len == -2) {
            return null;
        } else if (len == -1) {
            len = buff.getInt();
            int pos = buff.position();
            buff.position(buff.position() + len);
            return new String(buff.array(), pos, len, StandardCharsets.UTF_8);
        } else {
            int pos = buff.position();
            buff.position(buff.position() + len);
            return new String(buff.array(), pos, len, StandardCharsets.UTF_8);
        }
    }

Reply:
I have changed the code to read directly from the ByteBuffer instead of copying to an intermediate buffer, that is, to do this:
```
new String(buff.array(), pos, len, StandardCharsets.UTF_8);
```
However, the proposed solution of writing to the byte buffer the path segments separately just moves the work of splitting the string with the path into segments from the sort-batch thread into the transform thread, and it would complicate somewhat reading and writing to the ByteBuffer, which instead of having two strings for each entry (path and json) would now have n strings for the path and 1 for the json. Therefore, I will leave it as it is.